### PR TITLE
update java client: Add features to be used by the software-router

### PR DESCRIPTION
**Pull request by gvissers13**
So, in this pull request:
I had some classes... The purpose of this was to be able to use the new Java client to our software-router
https://tickets.cytomine.be/issue/CYTO-498

### CODE REVIEW

File: src/main/java/be/cytomine/client/models/Description.java [hunk 1]
```diff
@@ -42,6 +42,14 @@ public Description(String domain, Long idDomain, String description) {
         set("domainIdent",idDomain.toString());
         set("data", description);
     }
+    public Description addDescription(Long domainIdent, String domainClassName, String text) throws CytomineException
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/Description.java]: I don't get the need of this method. It's the role of the constructor to create a new object, see the code just above.
The old `Cytomine.addDescription(...)` method is replaced by `new Description(...).save()`

File: src/main/java/be/cytomine/client/models/ProcessingServer.java [hunk 2]
```diff
@@ -32,4 +36,9 @@ public String getDomainName() {
         return "processing_server";
     }
 
+    public void deleteProcessingServer(Long id) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/ProcessingServer.java]: It's the old way to do the job. Now we use the `delete()` method from the `Model` parent class

File: src/main/java/be/cytomine/client/models/Software.java [hunk 3]
```diff
@@ -29,4 +31,21 @@ public Software(String name, String serviceName, String resultType, String execu
         this.set("resultName", resultType);
         this.set("executeCommand", executeCommand);
     }
+    public Software deprecateSoftware(Long id) throws CytomineException {
+        Software software = this.fetch(id) ;
+        software.set("deprecated", true);
+        return update();
+    }
+
+    public Software addSoftware(String softwareVersion, String name, Long idSoftwareUserRepository, Long idDefaultProcessingServer, String resultType, String executeCommand, String pullingCommand) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/Software.java]: Old way to use the Java client. Merge with the constructor which has to be used to create new objects (first locally and then saved on the server with `save()` method)

File: src/main/java/be/cytomine/client/models/Software.java [hunk 4]
```diff
@@ -29,4 +31,21 @@ public Software(String name, String serviceName, String resultType, String execu
         this.set("resultName", resultType);
         this.set("executeCommand", executeCommand);
     }
+    public Software deprecateSoftware(Long id) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/Software.java]: It seems more meaningful to call the method `deprecate` 
```
Software mySoft = new Software(...)
mySoft.save()
mySoft.deprecate()
```

File: src/main/java/be/cytomine/client/models/SoftwareParameter.java [hunk 5]
```diff
@@ -44,4 +46,42 @@ public SoftwareParameter(String name, String type, Long idSoftware, String defau
     public SoftwareParameter(String name, String type, Long idSoftware, String defaultValue, boolean required, int index, boolean setByServer){
         this(name, type, idSoftware, defaultValue, required, index, null, null, null, setByServer);
     }
+
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/SoftwareParameter.java]: Idem as other. You don't implement the generic version of the client.

File: src/main/java/be/cytomine/client/models/SoftwareParameterConstraint.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import be.cytomine.client.CytomineException;
+
+public class SoftwareParameterConstraint extends Model<SoftwareParameterConstraint> {
+
+    public SoftwareParameterConstraint(){}
+    public SoftwareParameterConstraint(String value){
+        this.set("value", value);
+    }
+
+    public SoftwareParameterConstraint addSoftwareParameterConstraint(Long parameterConstraintId, Long softwareParameterId, String value) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/SoftwareParameterConstraint.java]: idem

File: src/main/java/be/cytomine/client/models/SoftwareUserRepository.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package be.cytomine.client.models;
+
+public class SoftwareUserRepository extends Model<SoftwareUserRepository> {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/SoftwareUserRepository.java:3]: That's the good way ! :+1:

File: src/main/java/be/cytomine/client/models/UploadedFile.java [hunk 8]
```diff
@@ -68,6 +70,14 @@ public UploadedFile(String originalFilename, String realFilename, String path, L
         }
     }
 
+    public AttachedFile uploadAttachedFile(String file, String domainClassName, Long domainIdent) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/UploadedFile.java]: AttachedFile is different from UploadedFile

File: src/main/java/be/cytomine/client/models/Job.java [hunk 9]
```diff
@@ -30,4 +54,23 @@ public Job(Long softwareId, Long projectId){
         set("project", projectId);
         set("software", softwareId);
     }
+
+    public Job addJob(Long softwareId, Long projectId) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/Job.java]: Not needed...

**Review by urubens — CHANGES_REQUESTED** [review]
Improve your PR because you don't get all the new philosophy behind this generic way of doing things. In fact, you implemented lots of things that are already dynamically done.

Compare your model's implementation with already existing ones.

Have you checked everything is working ?
Update the functional tests for models/collections you have modified/updated.

File: src/main/java/be/cytomine/client/models/Description.java (hunk 1 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/Description.java]: Done

File: src/main/java/be/cytomine/client/models/ProcessingServer.java (hunk 2 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/ProcessingServer.java]: done

File: src/main/java/be/cytomine/client/models/Software.java (hunk 3 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/Software.java]: done

File: src/main/java/be/cytomine/client/models/Software.java (hunk 4 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/Software.java]: done

File: src/main/java/be/cytomine/client/models/SoftwareParameter.java (hunk 5 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/SoftwareParameter.java]: done

File: src/main/java/be/cytomine/client/models/SoftwareParameterConstraint.java (hunk 6 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/SoftwareParameterConstraint.java]: done

File: src/main/java/be/cytomine/client/models/UploadedFile.java (hunk 8 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/UploadedFile.java]: done

File: src/main/java/be/cytomine/client/models/Job.java (hunk 9 above)
Comment (gvissers13) [reply src/main/java/be/cytomine/client/models/Job.java]: done

File: src/main/java/be/cytomine/client/models/UploadedFile.java [hunk 10]
```diff
@@ -70,12 +70,12 @@ public UploadedFile(String originalFilename, String realFilename, String path, L
         }
     }
 
-    public AttachedFile uploadAttachedFile(String file, String domainClassName, Long domainIdent) throws CytomineException {
+    public UploadedFile uploadAttachedFile(String file, String domainClassName, Long domainIdent) throws CytomineException {
```
Comment (urubens) [inline src/main/java/be/cytomine/client/models/UploadedFile.java]: Please pay attention to what you are doing instead of making tasks mechanically

Comment (urubens) [reply src/main/java/be/cytomine/client/models/UploadedFile.java]: Fix this and make something that makes sense

**Review by urubens — CHANGES_REQUESTED** [review]

**Review by urubens — APPROVED** [review]

**Review by urubens — APPROVED** [review]

**Pull request merged**